### DISCUSSION TITLE
Can we use OPA to Evaluate an Ansible Playbook

### DISCUSSION THREAD

**Discussion by sbalakrishn** [open-policy-agent/community#251] [OPA and Rego] [answered]
Hi All,
I would like to know if we can use OPA To evaluate Ansible Playbook, If we convert the Ansible PLaybook to a JSON file 
Can  someone help 

Regards
Sreejith Balakrishnan
Email : sreejith.balakrishnan@8x8.com

**Comment by anderseknert** [comment, marked as answer]
Hi @sbalakrishn 👋 Any structured data that can be represented as either JSON or YAML can be evaluated by OPA. If you could provide an example of what an ansible playbook looks like in JSON format, perhaps we could take it from there :)

**Reply by sbalakrishn** [reply]
Thanks Anderseknert  for replying...

I am looking to evaluate an Ansible Playbook that I have converted to a JSON file. 
Would we have a package that will help do this something like terraform.analysis for Ansible . Or can we just evaluate any JSON file just like that using OPA ?

The example of the JSON file that I want to evaluate is as below .

JSON FILE ...

[
  {
    "name": "Add or modify ansible.example.org A to 192.168.1.1\"",
    "community.general.nsupdate": {
      "key_name": "nsupdate",
      "key_secret": "+bFQtBCta7j2vWkjPkAFtgA==",
      "server": "10.1.1.1",
      "zone": "example.org",
      "record": "ansible",
      "value": "192.168.1.1"
    }
  },
  {
    "name": "Add or modify ansible.example.org A to 192.168.1.1, 192.168.1.2 and 192.168.1.3\"",
    "community.general.nsupdate": {
      "key_name": "nsupdate",
      "key_secret": "+bFQtBCta7j2vWkjPkAFtgA==",
      "server": "10.1.1.1",
      "zone": "example.org",
      "record": "ansible",
      "value": [
        "192.168.1.1",
        "192.168.1.2",
        "192.168.1.3"
      ]
    }
  },
  {
    "name": "Remove puppet.example.org CNAME",
    "community.general.nsupdate": {
      "key_name": "nsupdate",
      "key_secret": "+bFQtBCta7j2vWkjPkAFtgA==",
      "server": "10.1.1.1",
      "zone": "example.org",
      "record": "puppet",
      "type": "CNAME",
      "state": "absent"
    }
  },
  {
    "name": "Add 1.1.168.192.in-addr.arpa. PTR for ansible.example.org",
    "community.general.nsupdate": {
      "key_name": "nsupdate",
      "key_secret": "+bFQtBCta7j2vWkjPkAFtgA==",
      "server": "10.1.1.1",
      "record": "1.1.168.192.in-addr.arpa.",
      "type": "PTR",
      "value": "ansible.example.org.",
      "state": "present"
    }
  },
  {
    "name": "Remove 1.1.168.192.in-addr.arpa. PTR",
    "community.general.nsupdate": {
      "key_name": "nsupdate",
      "key_secret": "+bFQtBCta7j2vWkjPkAFtgA==",
      "server": "10.1.1.1",
      "record": "1.1.168.192.in-addr.arpa.",
      "type": "PTR",
      "state": "absent"
    }
  }
]

### **==The Policy file > policy.rego  == I am not able to fetch the Json value pair  > **"zone" :  "example.org"**** from the fole using the following combinations. Can you help . Below is my code for the rego policy

package package.json
# package ast
# package rules

import input as meteor
#########
# Policy
#########

# default allow_update := false
# allow_update {
#    #input.zone == "example.org"
#    #ansibleplan.zone == "example.org"
# }

# default allow_update := false
# allow_update {
#    m := ansibleplan.zone
#    m == "example.org"
# }

default allow_update := false
allow_update {
    # ansibleplan.community.general.nsupdate[_].zone == "example.org"
    meteor.community.general.nsupdate.zone == "example.org"
    # ansibleplan.name.community.general.nsupdate[_].zone == "example.org"
    # ansibleplan.name.community.general.nsupdate.zone == "example.org"
    # ansibleplan.name.zone == "example.org"
    # ansibleplan.name.community.general.nsupdate[_].zone == "example.org"
    # ansibleplan.name.zone == "example.org"
    # ansibleplan.community.general.nsupdate.zone == "example.org"
    # ansibleplan.community.general.nsupdate[_].zone[3] == "example.org"
}

**Reply by sbalakrishn** [reply]
=== I am currently Using the below policy code ===to fetch the correct line from the JSON file ..

meteor.json is the input file
policy-111.rego  is the policy file

./opa eval --format pretty --data policy-111.rego --input meteor.json "data.ast"

policy-111.rego  
----------------- 
package ast

import input as meteor

default allow_update := false
allow_update {
    meteor.community.general.nsupdate.zone == "example.org"
}

**Reply by anderseknert** [reply]
Hi again :) Without formatting, this is really hard to follow 😅 Could you perhaps provide an example in the [Rego Playground](https://play.openpolicyagent.org/)? You can use the Publish function to share a link to your policy, including the input data.

**Reply by sbalakrishn** [reply]
[](https://play.openpolicyagent.org/p/XXyXL8YP2v)

**Reply by sbalakrishn** [reply]
Link : https://play.openpolicyagent.org/p/XXyXL8YP2v

I have shared the Rego playground link above , kindly check and revert

Thank you

**Reply by sbalakrishn** [reply]
This must evaluate to "true" but it is evaluating to "false"
 -- Meaning it is not able to read the Key value pair from JSON as expected . Looks to be so

I am completely new to Rego and go lang

**Reply by anderseknert** [reply]
Something like should do it https://play.openpolicyagent.org/p/2s4nJlSeSi

**Reply by sbalakrishn** [reply]
Thank you Anders 

Thank you very much , That works like wonder

...wonderful

Have a great day :)

**Reply by anderseknert** [reply]
Great! You can mark the answer as accepted, so that others will see that it's "done" :)

**Reply by sbalakrishn** [reply]
Accepted !!!

**Reply by sbalakrishn** [reply]
> Something like should do it https://play.openpolicyagent.org/p/2s4nJlSeSi

Accepted
